### PR TITLE
config: add GrpcStream: gRPC mechanisms pulled out of GrpcMux

### PR DESCRIPTION
**Pull request by fredlas**
*Description*: Refactor necessary for #5466 to avoid completely duplicating this logic.
*Risk Level*: low
*Testing*: existing tests

### CODE REVIEW

File: source/common/config/grpc_mux_impl.cc [hunk 1]
```diff
@@ -84,17 +67,17 @@ void GrpcMuxImpl::queueDiscoveryRequest(const std::string& type_url) {
   drainRequests();
 }
 
-void GrpcMuxImpl::sendDiscoveryRequest(const std::string& type_url) {
-  if (stream_ == nullptr) {
+bool GrpcMuxImpl::sendDiscoveryRequest(const std::string& type_url) {
+  if (!discovery_grpc_stream_.available()) {
     ENVOY_LOG(debug, "No stream available to sendDiscoveryRequest for {}", type_url);
-    return;
+    return true; // TODO(fredlas) 'true' is the original behavior; should it be changed?
```
Comment (fredlas) [inline source/common/config/grpc_mux_impl.cc]: Having this case return true rather than false sticks to the existing behavior; the unit tests pass. However, it seems to me that the existing behavior is incorrect: a request that is attempted to be sent during a hiccup will be silently dropped, rather than left in the queue. I wanted to make sure I'm not misunderstanding what is intended here before I dive into changing all the unit tests to expect this behavior.

File: source/common/config/discovery_grpc_stream.h [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+namespace Envoy {
+namespace Config {
+
+// Oversees communication for gRPC-using xDS: reestablishes the gRPC channel when necessary, and
+// provides rate limiting of requests.
+template <class RequestProto, class ResponseProto>
+class DiscoveryGrpcStream : public Grpc::TypedAsyncStreamCallbacks<ResponseProto>,
+                            Logger::Loggable<Logger::Id::upstream> {
+public:
+  DiscoveryGrpcStream(Grpc::AsyncClientPtr async_client,
```
Comment (jmarantz) [inline source/common/config/discovery_grpc_stream.h]: WDYT of making these std::function args instead be pure virtual methods and you have to provide an override for those methods on instantiation rather than supplying them as lambdas? It looks like that might not be too bad at the call-site.

Comment (fredlas) [reply source/common/config/discovery_grpc_stream.h]: Done. Yeah, it turned out to be cleaner this way (assuming that decreasing use of multiple inheritance is not itself a goal).

File: source/common/config/discovery_grpc_stream.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+#include "envoy/grpc/async_client.h"
+
+#include "common/common/backoff_strategy.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+// Oversees communication for gRPC-using xDS: reestablishes the gRPC channel when necessary, and
```
Comment (htuch) [inline source/common/config/discovery_grpc_stream.h]: Nit "for gRPC xDS implementations (parent to both regular xDS and incremental xDS variants)."

File: source/common/config/discovery_grpc_stream.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+// provides rate limiting of requests.
+template <class RequestProto, class ResponseProto>
+class DiscoveryGrpcStream : public Grpc::TypedAsyncStreamCallbacks<ResponseProto>,
+                            public Logger::Loggable<Logger::Id::upstream> {
```
Comment (htuch) [inline source/common/config/discovery_grpc_stream.h]: I don't think you want to log upstream here, probably config?

File: source/common/config/grpc_mux_impl.cc [hunk 5]
```diff
@@ -40,40 +29,13 @@ GrpcMuxImpl::~GrpcMuxImpl() {
...
-  control_plane_stats_.connected_state_.set(1);
-  for (const auto type_url : subscriptions_) {
-    queueDiscoveryRequest(type_url);
-  }
-}
-
 void GrpcMuxImpl::drainRequests() {
   ENVOY_LOG(trace, "draining discovery requests {}", request_queue_.size());
-  while (!request_queue_.empty()) {
+  while (!request_queue_.empty() && checkRateLimitAllowsDrain(request_queue_.size())) {
```
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc]: Why couldn't this whole method move to the superclass? I'd prefer to avoid changing behavior during the simple refactor  and then do a separate PR for any goodness needed to support the on-demand case (with tests as needed).

File: source/common/config/grpc_mux_impl.cc (hunk 1 above)
Comment (htuch) [reply source/common/config/grpc_mux_impl.cc]: See other comment, would rather keep the refactor and behavioral change PR separate.

File: source/common/config/grpc_mux_impl.cc (hunk 5 above)
Comment (fredlas) [reply source/common/config/grpc_mux_impl.cc]: It can, and so could queueDiscoveryRequest; DiscoveryGrpcStream just needs to be templated on queue item type as well. I will do that.

Keeping the fix in a separate PR sounds good. However, I think you misunderstood me: I am not suggesting this change to support new things, rather, I think the logic is incorrect even for the existing vanilla xDS case. DiscoveryRequests can be silently dropped when the gRPC channel briefly goes down. That seems wrong to me, although I wanted to double-check that it isn't intended.

File: source/common/config/discovery_grpc_stream.h (hunk 4 above)
Comment (fredlas) [reply source/common/config/discovery_grpc_stream.h]: Thanks, didn't even notice that; done.

File: source/common/config/discovery_grpc_stream.h (hunk 3 above)
Comment (fredlas) [reply source/common/config/discovery_grpc_stream.h]: Nice, concise and more fully conveys the structure; thanks.

**Comment by fredlas** [comment]
I just realized that most of the names in this directory leave their association with xDS implicit, based on the directory they're living in. Would DiscoveryGrpcStream then make more sense as GrpcStream, despite sounding overly generic at first glance?

File: source/common/config/grpc_mux_impl.cc (hunk 5 above)
Comment (htuch) [reply source/common/config/grpc_mux_impl.cc]: I don't think this is an issue today; when the stream is re-established, a new DiscoverRequest will be sent, so any requests lost while disconnected don't matter.

**Comment by htuch** [comment]
Since it's in the `Config` namespace, abbreviating is fine. Even `XdsGrpcStream` would be shorter :)

File: source/common/config/grpc_mux_impl.h [hunk 6]
```diff
@@ -38,29 +38,16 @@ class GrpcMuxImpl : public GrpcMux,
   void pause(const std::string& type_url) override;
   void resume(const std::string& type_url) override;
 
-  // Grpc::AsyncStreamCallbacks
-  void onCreateInitialMetadata(Http::HeaderMap& metadata) override;
-  void onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) override;
-  void onReceiveMessage(std::unique_ptr<envoy::api::v2::DiscoveryResponse>&& message) override;
-  void onReceiveTrailingMetadata(Http::HeaderMapPtr&& metadata) override;
-  void onRemoteClose(Grpc::Status::GrpcStatus status, const std::string& message) override;
+  // queue_item is a resource type URL.
```
Comment (htuch) [inline source/common/config/grpc_mux_impl.h]: Prefer sticking with `type_url` here.

**Comment by fredlas** [comment]
Forgot to tag this as part of work on #4991.
